File: figma-renders/parse_figma.py
import json
import sys
import os

def rgba_to_hex(c, opacity=1.0):
    r = int(c.get('r', 0) * 255)
    g = int(c.get('g', 0) * 255)
    b = int(c.get('b', 0) * 255)
    a = c.get('a', 1.0)
    return f"#{r:02X}{g:02X}{b:02X}", a * opacity

def extract_node_tree(node, depth=0, results=None):
    if results is None:
        results = []

    indent = "  " * depth
    name = node.get('name', 'unnamed')
    ntype = node.get('type', 'unknown')
    bbox = node.get('absoluteBoundingBox', {})
    w = bbox.get('width', 0)
    h = bbox.get('height', 0)
    x = bbox.get('x', 0)
    y = bbox.get('y', 0)

    line = f"{indent}{ntype}: \"{name}\" ({w}x{h}) at ({x},{y})"
    results.append(line)

    # Fills
    fills = node.get('fills', [])
    for f in fills:
        if f.get('visible', True) is not False:
            ftype = f.get('type')
            if ftype == 'SOLID':
                c = f.get('color', {})
                hex_color, alpha = rgba_to_hex(c, f.get('opacity', 1.0))
                results.append(f"{indent}  Fill: {hex_color} alpha={alpha:.2f}")
            elif ftype == 'GRADIENT_LINEAR':
                stops = f.get('gradientStops', [])
                results.append(f"{indent}  Gradient Linear: {len(stops)} stops")
                for s in stops:
                    c = s.get('color', {})
                    hex_color, _ = rgba_to_hex(c)
                    results.append(f"{indent}    Stop {s.get('position',0):.2f}: {hex_color} a={c.get('a',1):.2f}")
                handles = f.get('gradientHandlePositions', [])
                if handles:
                    results.append(f"{indent}    Handles: {handles}")
            elif ftype == 'GRADIENT_RADIAL':
                stops = f.get('gradientStops', [])
                results.append(f"{indent}  Gradient Radial: {len(stops)} stops")
                for s in stops:
                    c = s.get('color', {})
                    hex_color, _ = rgba_to_hex(c)
                    results.append(f"{indent}    Stop {s.get('position',0):.2f}: {hex_color} a={c.get('a',1):.2f}")
            elif ftype == 'IMAGE':
                results.append(f"{indent}  Fill: IMAGE ref={f.get('imageRef','')}")

    # Effects
    effects = node.get('effects', [])
    for e in effects:
        if e.get('visible', True):
            c = e.get('color', {})
            hex_color, _ = rgba_to_hex(c) if c else ('#000000', 1)
            offset = e.get('offset', {})
            results.append(f"{indent}  Effect: {e.get('type')} color={hex_color} a={c.get('a',0):.3f} offset=({offset.get('x',0)},{offset.get('y',0)}) blur={e.get('radius',0)} spread={e.get('spread',0)}")

    # Corner radius
    cr = node.get('cornerRadius')
    rcr = node.get('rectangleCornerRadii')
    if cr:
        results.append(f"{indent}  BorderRadius: {cr}")
    if rcr and rcr != [0,0,0,0]:
        results.append(f"{indent}  CornerRadii: {rcr}")

    # Text style
    style = node.get('style', {})
    if style and ntype == 'TEXT':
        chars = node.get('characters', '')
        results.append(f"{indent}  Text: \"{chars}\"")
        results.append(f"{indent}  Font: {style.get('fontFamily','')} weight={style.get('fontWeight','')} size={style.get('fontSize','')}px")
        results.append(f"{indent}  LineHeight: {style.get('lineHeightPx','')}px letterSpacing={style.get('letterSpacing','')} align={style.get('textAlignHorizontal','')}")
        results.append(f"{indent}  TextCase: {style.get('textCase','ORIGINAL')} Decoration: {style.get('textDecoration','NONE')}")

    # Layout
    lm = node.get('layoutMode')
    if lm:
        isp = node.get('itemSpacing', 0)
        pl = node.get('paddingLeft', 0)
        pr = node.get('paddingRight', 0)
        pt = node.get('paddingTop', 0)
        pb = node.get('paddingBottom', 0)
        caa = node.get('counterAxisAlignItems', '')
        paa = node.get('primaryAxisAlignItems', '')
        results.append(f"{indent}  Layout: {lm} gap={isp} align={caa}/{paa}")
        if any([pl, pr, pt, pb]):
            results.append(f"{indent}  Padding: L={pl} R={pr} T={pt} B={pb}")

    # Strokes
    strokes = node.get('strokes', [])
    for s in strokes:
        if s.get('visible', True) is not False:
            c = s.get('color', {})
            if c:
                hex_color, _ = rgba_to_hex(c)
                sw = node.get('strokeWeight', 1)
                results.append(f"{indent}  Stroke: {hex_color} weight={sw}")

    # Boolean operation
    bo = node.get('booleanOperation')
    if bo:
        results.append(f"{indent}  BooleanOp: {bo}")

    # Fill geometry (for clip paths)
    fg = node.get('fillGeometry', [])
    if fg and ntype in ('BOOLEAN_OPERATION', 'VECTOR'):
        for g in fg:
            path = g.get('path', '')
            if len(path) > 200:
                results.append(f"{indent}  FillGeometry: {path[:200]}...")
            else:
                results.append(f"{indent}  FillGeometry: {path}")

    # Clips content
    if node.get('clipsContent'):
        results.append(f"{indent}  ClipsContent: true")

    # Children
    for child in node.get('children', []):
        extract_node_tree(child, depth + 1, results)

    return results

def process_section(filename, node_id, section_name):
    with open(filename, 'r') as f:
        data = json.load(f)

    node = data['nodes'][node_id]['document']
    results = [f"\n{'='*80}", f"SECTION: {section_name} (node {node_id})", f"{'='*80}"]
    extract_node_tree(node, 0, results)
    return "\n".join(results)

sections = [
    ('header_raw.json', '236:410', 'Header'),
    ('hero_raw.json', '236:331', 'Hero'),
    ('services_raw.json', '236:434', 'Services+Projects'),
    ('team_raw.json', '236:472', 'Team/MeetExperts'),
    ('cta_raw.json', '236:522', 'CTA/Contact'),
    ('footer_raw.json', '236:577', 'Footer'),
]

base_dir = r'D:\Seyon\react-app\figma-renders'
all_output = []

for filename, node_id, section_name in sections:
    filepath = os.path.join(base_dir, filename)
    try:
        output = process_section(filepath, node_id, section_name)
        all_output.append(output)
    except Exception as e:
        all_output.append(f"\nERROR processing {section_name}: {e}")

with open(os.path.join(base_dir, 'parsed_specs.txt'), 'w', encoding='utf-8') as f:
    f.write("\n".join(all_output))

print("Done! Output saved to parsed_specs.txt")
print(f"Total lines: {sum(len(o.split(chr(10))) for o in all_output)}")

File: figma-renders/parse_figma.js
const fs = require('fs');
const path = require('path');

const baseDir = 'D:/Seyon/react-app/figma-renders';

function rgbaToHex(c, opacity = 1.0) {
  const r = Math.round((c.r || 0) * 255);
  const g = Math.round((c.g || 0) * 255);
  const b = Math.round((c.b || 0) * 255);
  const hex = '#' + [r, g, b].map(v => v.toString(16).padStart(2, '0').toUpperCase()).join('');
  return { hex, alpha: (c.a || 1) * opacity };
}

function extractNodeTree(node, depth = 0, results = []) {
  const indent = '  '.repeat(depth);
  const name = node.name || 'unnamed';
  const ntype = node.type || 'unknown';
  const bbox = node.absoluteBoundingBox || {};
  const w = bbox.width || 0;
  const h = bbox.height || 0;
  const x = bbox.x || 0;
  const y = bbox.y || 0;

  results.push(`${indent}${ntype}: "${name}" (${w}x${h}) at (${x},${y})`);

  // Fills
  const fills = node.fills || [];
  for (const f of fills) {
    if (f.visible !== false) {
      if (f.type === 'SOLID') {
        const { hex, alpha } = rgbaToHex(f.color || {}, f.opacity != null ? f.opacity : 1);
        results.push(`${indent}  Fill: ${hex} alpha=${alpha.toFixed(2)}`);
      } else if (f.type === 'GRADIENT_LINEAR') {
        const stops = f.gradientStops || [];
        results.push(`${indent}  Gradient Linear: ${stops.length} stops`);
        for (const s of stops) {
          const { hex } = rgbaToHex(s.color || {});
          results.push(`${indent}    Stop ${(s.position || 0).toFixed(2)}: ${hex} a=${(s.color?.a || 1).toFixed(2)}`);
        }
        if (f.gradientHandlePositions) {
          results.push(`${indent}    Handles: ${JSON.stringify(f.gradientHandlePositions)}`);
        }
      } else if (f.type === 'GRADIENT_RADIAL') {
        const stops = f.gradientStops || [];
        results.push(`${indent}  Gradient Radial: ${stops.length} stops`);
        for (const s of stops) {
          const { hex } = rgbaToHex(s.color || {});
          results.push(`${indent}    Stop ${(s.position || 0).toFixed(2)}: ${hex} a=${(s.color?.a || 1).toFixed(2)}`);
        }
      } else if (f.type === 'IMAGE') {
        results.push(`${indent}  Fill: IMAGE ref=${f.imageRef || ''} scaleMode=${f.scaleMode || ''}`);
      }
    }
  }

  // Effects
  const effects = node.effects || [];
  for (const e of effects) {
    if (e.visible !== false) {
      const c = e.color || {};
      const { hex } = rgbaToHex(c);
      const offset = e.offset || {};
      results.push(`${indent}  Effect: ${e.type} color=${hex} a=${(c.a || 0).toFixed(3)} offset=(${offset.x || 0},${offset.y || 0}) blur=${e.radius || 0} spread=${e.spread || 0}`);
    }
  }

  // Corner radius
  if (node.cornerRadius) {
    results.push(`${indent}  BorderRadius: ${node.cornerRadius}`);
  }
  if (node.rectangleCornerRadii && JSON.stringify(node.rectangleCornerRadii) !== '[0,0,0,0]') {
    results.push(`${indent}  CornerRadii: [${node.rectangleCornerRadii.join(',')}]`);
  }

  // Text style
  const style = node.style || {};
  if (ntype === 'TEXT' && Object.keys(style).length > 0) {
    results.push(`${indent}  Text: "${node.characters || ''}"`);
    results.push(`${indent}  Font: ${style.fontFamily || ''} weight=${style.fontWeight || ''} size=${style.fontSize || ''}px`);
    results.push(`${indent}  LineHeight: ${style.lineHeightPx || ''}px letterSpacing=${style.letterSpacing || 0} align=${style.textAlignHorizontal || ''}`);
    results.push(`${indent}  TextCase: ${style.textCase || 'ORIGINAL'} Decoration: ${style.textDecoration || 'NONE'}`);
  }

  // Layout
  if (node.layoutMode) {
    const isp = node.itemSpacing || 0;
    const pl = node.paddingLeft || 0;
    const pr = node.paddingRight || 0;
    const pt = node.paddingTop || 0;
    const pb = node.paddingBottom || 0;
    results.push(`${indent}  Layout: ${node.layoutMode} gap=${isp} counterAxis=${node.counterAxisAlignItems || ''} primaryAxis=${node.primaryAxisAlignItems || ''}`);
    if (pl || pr || pt || pb) {
      results.push(`${indent}  Padding: L=${pl} R=${pr} T=${pt} B=${pb}`);
    }
  }

  // Strokes
  const strokes = node.strokes || [];
  for (const s of strokes) {
    if (s.visible !== false && s.color) {
      const { hex } = rgbaToHex(s.color);
      results.push(`${indent}  Stroke: ${hex} weight=${node.strokeWeight || 1} align=${node.strokeAlign || ''}`);
    }
  }

  // Boolean operation
  if (node.booleanOperation) {
    results.push(`${indent}  BooleanOp: ${node.booleanOperation}`);
  }

  // Fill geometry (clip paths)
  if (node.fillGeometry && (ntype === 'BOOLEAN_OPERATION' || ntype === 'VECTOR')) {
    for (const g of node.fillGeometry) {
      const p = g.path || '';
      if (p.length > 300) {
        results.push(`${indent}  FillGeometry: ${p.substring(0, 300)}...`);
      } else {
        results.push(`${indent}  FillGeometry: ${p}`);
      }
    }
  }

  // Clips content
  if (node.clipsContent) {
    results.push(`${indent}  ClipsContent: true`);
  }

  // Opacity
  if (node.opacity != null && node.opacity !== 1) {
    results.push(`${indent}  Opacity: ${node.opacity}`);
  }

  // Children
  for (const child of (node.children || [])) {
    extractNodeTree(child, depth + 1, results);
  }

  return results;
}

function processSection(filename, nodeId, sectionName) {
  const filepath = path.join(baseDir, filename);
  const data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
  const node = data.nodes[nodeId].document;
  const results = [
    '',
    '='.repeat(80),
    `SECTION: ${sectionName} (node ${nodeId})`,
    '='.repeat(80),
  ];
  extractNodeTree(node, 0, results);
  return results.join('\n');
}

const sections = [
  ['header_raw.json', '236:410', 'Header'],
  ['hero_raw.json', '236:331', 'Hero'],
  ['services_raw.json', '236:434', 'Services+Projects'],
  ['team_raw.json', '236:472', 'Team/MeetExperts'],
  ['cta_raw.json', '236:522', 'CTA/Contact'],
  ['footer_raw.json', '236:577', 'Footer'],
];

let allOutput = [];
for (const [filename, nodeId, sectionName] of sections) {
  try {
    const output = processSection(filename, nodeId, sectionName);
    allOutput.push(output);
  } catch (e) {
    allOutput.push(`\nERROR processing ${sectionName}: ${e.message}`);
  }
}

const outPath = path.join(baseDir, 'parsed_specs.txt');
fs.writeFileSync(outPath, allOutput.join('\n'), 'utf-8');
console.log('Done! Output saved to parsed_specs.txt');
console.log('Total chars:', allOutput.join('\n').length);
